base/setup/lib/utils/partlist.c

```diff
@@ -835,6 +825,145 @@ InitializePartitionEntry(
+     *the output buffer to IOCTL_MOUNTDEV_QUERY_DEVICE_NAME.
+     * It's based on MOUNTDEV_NAME (mountmgr.h).
+     * Doing it this way prevents memory allocation.
+     * The device name won't be longer.
+     */
+    struct
+    {
+        USHORT NameLength;
+        WCHAR Name[256];
+    } DeviceName;
+
+    /* If we already have a volume device name, do nothing more */
+    if (*Volume->Info.DeviceName)
+        return STATUS_SUCCESS;
+
+    if (!AltDeviceName)
+    {
+        PPARTENTRY PartEntry;
+        PartEntry = Volume->PartEntry;
+        ASSERT(PartEntry);
+        ASSERT(PartEntry->IsPartitioned && PartEntry->PartitionNumber != 0);
+        AltDeviceName = PartEntry->DeviceName;
+    }
+
+    /* Make a temporary device name for the volume */
+    Status = RtlStringCchCopyW(Volume->Info.DeviceName,
+                               _countof(Volume->Info.DeviceName),
+                               AltDeviceName);
+    ASSERT(NT_SUCCESS(Status));
+
+    /* Try to open the volume (if it is valid, this will also mount it) */
+    Status = pOpenDevice(Volume->Info.DeviceName, &VolumeHandle);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("pOpenDevice() failed, Status 0x%08lx\n", Status);
+        return Status;
+    }
+
+    /* Retrieve the non-persistent volume device name */
+    Status = NtDeviceIoControlFile(VolumeHandle,
+                                   NULL, NULL, NULL,
+                                   &IoStatusBlock,
+                                   IOCTL_MOUNTDEV_QUERY_DEVICE_NAME,
+                                   NULL, 0,
+                                   &DeviceName, sizeof(DeviceName));
+    NtClose(VolumeHandle);
+
+    // NOTE: If a memory allocation were needed, Status would be
+    // equal to STATUS_BUFFER_OVERFLOW, and one would allocate
+    // a buffer of size
+    //     FIELD_OFFSET(MOUNTDEV_NAME, Name[0]) + DeviceName.NameLength
+    // before calling the IOCTL again on the new buffer and size.
+
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("IOCTL_MOUNTDEV_QUERY_DEVICE_NAME failed, Status 0x%08lx\n", Status);
+        return Status;
+    }
+
+    /* Copy the volume device name */
+    Status = RtlStringCchCopyNW(Volume->Info.DeviceName,
+                                _countof(Volume->Info.DeviceName),
+                                DeviceName.Name,
+                                DeviceName.NameLength / sizeof(WCHAR));
+    ASSERT(NT_SUCCESS(Status));
+    return Status;
+}
+
+// Old name: CreateVolume
+static
+PVOLENTRY
+InitVolume(
+    _In_ PPARTLIST List,
+    _In_opt_ PPARTENTRY PartEntry)
+{
+    PVOLENTRY Volume;
+
+    Volume = RtlAllocateHeap(ProcessHeap,
+                             HEAP_ZERO_MEMORY,
+                             sizeof(VOLENTRY));
+    if (!Volume)
+        return NULL;
+
+    /* Reset some volume information */
+
+    /* No device name for now */
+    Volume->Info.DeviceName[0] = L'\0';
+    // Volume->Info.VolumeName[0] = L'\0';
```

Comment:
```suggestion
```
forgotten to remove? Or prepared for the future?

Reply:
Prepared for the future